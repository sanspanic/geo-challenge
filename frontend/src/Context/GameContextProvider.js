import React, { useState, useEffect } from "react";
import GameContext from "./GameContext";
import geoApi from "../API/geoAPI";

const GameContextProvider = ({ children }) => {
  const [speedBonus, setSpeedBonus] = useState(0);
  const [width, setWidth] = useState(100);
  const [level, setLevel] = useState(1);
  const [score, setScore] = useState(0);
  const [countries, setCountries] = useState([]);
  const [rank, setRank] = useState(1);
  const [hasLoaded, setHasLoaded] = useState(false);
  const [lives, setLives] = useState(5);
  const [finalScore, setFinalScore] = useState(0);

  const [status, setStatus] = useState({
    isActive: false,
    isLost: true,
    isWon: false,
  });

  useEffect(() => {
    const getCountryData = async () => {
      try {
        const res = await geoApi.getData();
        setCountries(res);
        setHasLoaded(true);
      } catch (err) {
        console.log(err);
      }
    };
    getCountryData();
  }, []);

  return (
    <GameContext.Provider
      value={{
        level,
        setLevel,
        score,
        setScore,
        countries,
        setCountries,
        hasLoaded,
        status,
        setStatus,
        lives,
        setLives,
        width,
        setWidth,
        speedBonus,
        setSpeedBonus,
        rank,
        setRank,
        finalScore,
        setFinalScore,
      }}
    >
      {children}
    </GameContext.Provider>
  );
};

export default GameContextProvider;
