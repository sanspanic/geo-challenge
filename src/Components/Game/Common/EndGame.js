import React, { useContext, useEffect, useState } from "react";
import GameContext from "../../../Context/GameContext";
import { Link } from "react-router-dom";
import Loser from "../../../Assets/loser.jpg";
import LostInForest from "../../../Assets/LostInForest.jpg";
import GoogleMaps from "../../../Assets/GoogleMaps.jpg";
import Inventor from "../../../Assets/Inventor.jpg";
import Cartographer from "../../../Assets/Cartographer.jpg";
import Compass from "../../../Assets/compass.jpg";
import Ship from "../../../Assets/Ship.jpg";
import SatNav from "../../../Assets/SatNav.jpg";
import { getPercentile, ranks } from "../Common/helpers";
import { Trophy } from "phosphor-react";
import backendAPI from "../../../API/backendAPI";

const EndGame = () => {
  const {
    setStatus,
    status,
    score,
    setScore,
    setMistakes,
    setLevel,
    mistakes,
    speedBonus,
  } = useContext(GameContext);
  const [rank, setRank] = useState(1);
  const [imgSrc, setImgSrc] = useState("");
  const [newHighScore, setNewHighScore] = useState(0);

  //update highscore
  useEffect(() => {
    setScore(1207);
    const tryUpdateScore = async () => {
      try {
        const res = await backendAPI.updateHighscore(
          JSON.parse(localStorage.getItem("user")).username,
          { score }
        );
        console.log(res);
        if (res.highscore) {
          setNewHighScore(res.highscore);
        }
      } catch (err) {
        console.log(err);
      }
    };
    tryUpdateScore();
  }, [score]);

  useEffect(() => {
    const percentile = getPercentile(score);
    if (percentile > (100 / 8) * 7) {
      setRank(8);
      setImgSrc(GoogleMaps);
    } else if ((100 / 8) * 7 >= percentile && percentile > (100 / 8) * 6) {
      setRank(7);
      setImgSrc(Inventor);
    } else if ((100 / 8) * 6 >= percentile && percentile > (100 / 8) * 5) {
      setRank(6);
      setImgSrc(Cartographer);
    } else if ((100 / 8) * 5 >= percentile && percentile > (100 / 8) * 4) {
      setRank(5);
      setImgSrc(Compass);
    } else if ((100 / 8) * 4 >= percentile && percentile > (100 / 8) * 3) {
      setRank(4);
      setImgSrc(Ship);
    } else if ((100 / 8) * 3 >= percentile && percentile > (100 / 8) * 2) {
      setRank(3);
      setImgSrc(SatNav);
    } else if ((100 / 8) * 2 >= percentile && percentile > 100 / 8) {
      setRank(2);
      setImgSrc(LostInForest);
    } else if (percentile <= 100 / 8) {
      setRank(1);
      setImgSrc(Loser);
    }
  }, [score, setRank]);

  const restart = () => {
    setLevel(1);
    setStatus({ isActive: false, isWon: false, isLost: false });
    setScore(0);
    setMistakes(5);
  };

  return (
    <div className="grid grid-cols-2 gap-10">
      <div className="rounded flex flex-col items-center bg-white bg-opacity-50 border p-10 max-w-screen-sm md:max-w-screen-md">
        <h2 className="font-black text-4xl">Game Over!</h2>
        <ul className="endgame my-10 w-full">
          <h3 className="text-center font-bold border-b border-black p-1">
            Final Tally
          </h3>
          <li className="flex justify-between px-3 py-1">
            <span>Speed bonus: </span>
            <span>
              {speedBonus / 50} x 50 = {speedBonus}
            </span>
          </li>
          <li className="flex justify-between px-3 py-1">
            <span>Accuracy bonus:Â </span>
            <span>
              {mistakes} x 50 = {mistakes * 50}
            </span>
          </li>
          <li className="flex justify-between px-3 py-1">
            <span>Final score: </span>
            <span className="font-bold">{score}</span>
          </li>
          <li className="flex justify-between px-3 py-1 text-cerise-500 font-bold">
            Final rank: <span>{ranks[0][rank].name}</span>
          </li>
          <li className="flex justify-between px-3 py-1 font-bold">
            Next rank:{" "}
            {rank === 8 ? (
              <span>
                You're already the best <Trophy className="inline" size={28} />
              </span>
            ) : (
              <span>{ranks[0][rank + 1].name}</span>
            )}
          </li>
        </ul>
        {newHighScore === 0 ? null : (
          <p className="text-cerise-500">NEW HIGHSCORE!</p>
        )}
        <p className="text-gray-600 pb-5">
          Did you make it to the{" "}
          <Link to="leaderboard" className="underline italic text-black">
            Leaderboard
          </Link>
          ?
        </p>
        <button
          onClick={restart}
          className="px-4 py-2 bg-gradient-green rounded font-bold text-white transition duration-400 hover:text-black transform-gpu hover:scale-110"
        >
          Play again!
        </button>
      </div>
      <div className="rank-card bg-white bg-opacity-50 border self-center p-10 rounded flex flex-col items-center self-stretch">
        <h2 className="text-4xl font-black text-center">
          Your Rank:
          <span className="text-cerise-500"> {ranks[0][rank].name}</span>
        </h2>
        <img className="py-5" src={imgSrc}></img>
        <p className="max-w-sm">{ranks[0][rank].description}</p>
      </div>
    </div>
  );
};

export default EndGame;
