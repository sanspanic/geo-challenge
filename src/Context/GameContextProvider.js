import React, { useState } from "react";
import GameContext from "./GameContext";

const GameContextProvider = ({ children }) => {
  const [level, setLevel] = useState({ num: 1, title: "Guess The Flag" });
  const [score, setScore] = useState(0);

  return (
    <GameContext.Provider value={{ level, setLevel, score, setScore }}>
      {children}
    </GameContext.Provider>
  );
};

export default GameContextProvider;
