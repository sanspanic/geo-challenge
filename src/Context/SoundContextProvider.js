import React, { useState } from "react";
import SoundContext from "./SoundContext";
import useSound from "use-sound";
import correct from "../Assets/sounds/correct.wav";
import wrong from "../Assets/sounds/wrong.wav";
import levelup from "../Assets/sounds/levelup.wav";
import gamelost from "../Assets/sounds/gamelost.wav";
import gamewon from "../Assets/sounds/gamewon.wav";

const SoundContextProvider = ({ children }) => {
  const [playCorrect] = useSound(correct, { volume: 0.25 });
  const [playWrong] = useSound(wrong, { volume: 0.25 });
  const [playLevelUp] = useSound(levelup, { volume: 0.25 });
  const [playGameOver] = useSound(gamelost, { volume: 0.25 });
  const [playGameWon] = useSound(gamewon, { volume: 0.55 });

  return (
    <SoundContext.Provider
      value={{ playCorrect, playWrong, playLevelUp, playGameOver, playGameWon }}
    >
      {children}
    </SoundContext.Provider>
  );
};

export default SoundContextProvider;
