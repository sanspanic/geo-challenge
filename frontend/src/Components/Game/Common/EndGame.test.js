import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import GameContextProvider from "../../../Context/GameContextProvider";
import SoundContextProvider from "../../../Context/SoundContextProvider";
import EndGame from "./EndGame";

test("renders without crashing", () => {
  render(
    <GameContextProvider>
      <SoundContextProvider>
        <MemoryRouter>
          <EndGame />
        </MemoryRouter>
      </SoundContextProvider>
    </GameContextProvider>
  );
});

test("matches snapshot", () => {
  const { asFragment } = render(
    <GameContextProvider>
      <SoundContextProvider>
        <MemoryRouter>
          <EndGame />
        </MemoryRouter>
      </SoundContextProvider>
    </GameContextProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});
