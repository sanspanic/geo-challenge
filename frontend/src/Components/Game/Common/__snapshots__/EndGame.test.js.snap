// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid sm:grid-cols-2 my-5 gap-10 w-11/12"
  >
    <div
      class=" shadow-xl rounded flex flex-col items-center bg-white bg-opacity-50 border p-10 max-w-screen-sm md:max-w-screen-md"
    >
      <h2
        class="font-black text-4xl"
      >
        Game Over!
      </h2>
      <p
        class="text-sm mt-5"
      >
        Do you want to play all levels? 
        <a
          class="italic underline"
          href="/register"
        >
          Register
        </a>
        .
      </p>
      <ul
        class="font-display endgame my-10 w-full"
      >
        <h3
          class="text-center font-bold border-b border-black p-1"
        >
          Final Tally
        </h3>
        <li
          class="flex justify-between px-3 py-1"
        >
          <span>
            Base score: 
          </span>
          <span>
            0
          </span>
        </li>
        <li
          class="flex justify-between px-3 py-1"
        >
          <span>
            Speed bonus: 
          </span>
          <span>
            0 x 50 = 0
          </span>
        </li>
        <li
          class="flex justify-between px-3 py-1"
        >
          <span>
            Accuracy bonus:Â 
          </span>
          <span>
            5 x 50 = 250
          </span>
        </li>
        <li
          class="flex justify-between px-3 py-1"
        >
          <span>
            Final score:
          </span>
          <span
            class="font-bold"
          >
            250
          </span>
        </li>
        <li
          class="flex justify-between px-3 py-1 text-cerise-500 font-bold"
        >
          Final rank: 
          <span>
            Loser
          </span>
        </li>
        <li
          class="flex justify-between px-3 py-1 font-bold"
        >
          Next rank: 
          <span>
            Lost in Forest
          </span>
        </li>
      </ul>
      <p
        class="text-gray-600 pb-5"
      >
        Did you make it to the 
        <a
          class="underline italic text-black"
          href="/leaderboard"
        >
          Leaderboard
        </a>
        ?
      </p>
      <button
        class="px-4 py-2 bg-gradient-green rounded font-bold text-white transition duration-400 hover:text-black transform-gpu hover:scale-110"
      >
        Play again!
      </button>
    </div>
    <div
      class="mesh shadow-xl rank-card bg-white bg-opacity-50 border self-center p-10 rounded flex flex-col items-center self-stretch"
    >
      <h2
        class="text-4xl font-black text-center"
      >
        Your Rank:
        <span
          class="text-cerise-500"
        >
           Loser
        </span>
      </h2>
      <img
        alt="rank"
        class="py-5"
        src="loser.jpg"
      />
      <p
        class="max-w-sm font-display"
      >
        Geographically speaking, you are not the sharpest tool in the shed. At
    any given moment in time, you are likely to be lost and actively
    searching for an exit. Life is difficult.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
